File: p11-platform/apps/web/utils/propertyaudit/claude-connector.ts
/**
 * Claude Connector for PropertyAudit
 * Uses Anthropic API for GEO audits
 */

import Anthropic from '@anthropic-ai/sdk'
import { 
  AnswerBlockSchema, 
  type AnswerBlock, 
  type Connector, 
  type ConnectorContext, 
  type ConnectorResult,
  getGeoConfig 
} from './types'
import { performWebSearch, formatSearchResultsForLLM } from './web-search'

function buildPrompt(context: ConnectorContext): string {
  const domains = context.brandDomains.join(', ')
  const competitors = context.competitors.join(', ')
  const location = context.propertyLocation
  
  const lines = [
    `Task: Perform a GEO audit for this specific real estate property and return ONLY valid JSON matching the exact schema.`,
  ]

  // Add property location context if available (prevents hallucinations)
  if (location && location.city && location.state) {
    lines.push(``)
    lines.push(`Property Details:`)
    lines.push(`- Name: ${context.brandName}`)
    lines.push(`- Location: ${location.city}, ${location.state}`)
    if (location.fullAddress) {
      lines.push(`- Address: ${location.fullAddress}`)
    }
    if (location.websiteUrl) {
      lines.push(`- Official Website: ${location.websiteUrl}`)
    }
    lines.push(``)
    lines.push(`CRITICAL: This property is located in ${location.city}, ${location.state}.`)
    lines.push(`Do NOT confuse with other AMLI properties in different cities (e.g., Denver, Austin, Chicago).`)
    lines.push(`All information must be specific to the ${location.city}, ${location.state} location.`)
    lines.push(`If you cite URLs, ensure they reference the ${location.city} property, not other locations.`)
  }

  lines.push(``)
  lines.push(`Query: ${context.queryText}`)
  lines.push(`Brand: ${context.brandName}`)
  lines.push(`Brand domains: ${domains || '—'}`)
  lines.push(`Competitors: ${competitors || '—'}`)
  lines.push(``)
  lines.push(`Requirements:`)
  lines.push(`- Produce an ordered list of providers/brands relevant to the query (name, domain, rationale, position starting at 1).`)
  lines.push(`- Include citations with absolute URLs and their domains.`)
  lines.push(`- Summarize the answer in 1-2 sentences.`)
  lines.push(`- If no grounded sources are available, set notes.flags to include "no_sources".`)
  lines.push(``)
  lines.push(`Output format - Return ONLY raw JSON (no markdown code blocks, no explanations, no text before or after):`)
  lines.push(`{`)
  lines.push(`  "ordered_entities": [`)
  lines.push(`    {"name": "...", "domain": "...", "rationale": "...", "position": 1}`)
  lines.push(`  ],`)
  lines.push(`  "citations": [`)
  lines.push(`    {"url": "...", "domain": "...", "entity_ref": "1"}`)
  lines.push(`  ],`)
  lines.push(`  "answer_summary": "...",`)
  lines.push(`  "notes": {"flags": []}`)
  lines.push(`}`)
  
  return lines.join('\n')
}

function tryParseJson(content: string): unknown {
  // First, try direct JSON parse
  try {
    return JSON.parse(content)
  } catch {
    // Continue to fallbacks
  }

  // Strip markdown code blocks
  let cleaned = content.trim()
  cleaned = cleaned.replace(/^```(?:json)?\s*\n?/i, '')
  cleaned = cleaned.replace(/\n?```\s*$/i, '')
  cleaned = cleaned.trim()

  try {
    return JSON.parse(cleaned)
  } catch {
    // Continue to fallbacks
  }

  // Extract JSON object from text
  const firstBrace = cleaned.indexOf('{')
  const lastBrace = cleaned.lastIndexOf('}')

  if (firstBrace !== -1 && lastBrace !== -1 && lastBrace > firstBrace) {
    const jsonCandidate = cleaned.substring(firstBrace, lastBrace + 1)
    try {
      return JSON.parse(jsonCandidate)
    } catch {
      // Continue to fallbacks
    }
  }

  // Fallback: regex match from end
  const match = content.match(/\{[\s\S]*\}$/)
  if (match) {
    try {
      return JSON.parse(match[0])
    } catch {
      // All parsing attempts failed
    }
  }

  return null
}

function coerceToAnswerBlock(candidate: unknown): AnswerBlock | null {
  const direct = AnswerBlockSchema.safeParse(candidate)
  if (direct.success) {
    return direct.data
  }

  if (!candidate || typeof candidate !== 'object') {
    return null
  }

  const obj = candidate as Record<string, unknown>

  // Try to extract ordered_entities from various formats
  const entitiesSource = Array.isArray(obj.ordered_entities)
    ? obj.ordered_entities
    : Array.isArray(obj.results)
    ? obj.results
    : Array.isArray(obj.providers)
    ? obj.providers
    : null

  const orderedEntities: AnswerBlock['ordered_entities'] = []
  
  if (entitiesSource) {
    for (let i = 0; i < entitiesSource.length; i++) {
      const item = entitiesSource[i]
      if (!item || typeof item !== 'object') continue
      
      const ent = item as Record<string, unknown>
      const name = typeof ent.name === 'string' ? ent.name : null
      const domain = typeof ent.domain === 'string' ? ent.domain : null
      
      if (!name || !domain) continue
      
      orderedEntities.push({
        name,
        domain,
        rationale: typeof ent.rationale === 'string' ? ent.rationale : 'No rationale provided.',
        position: typeof ent.position === 'number' ? ent.position : i + 1
      })
    }
  }

  // Extract citations
  const citations: AnswerBlock['citations'] = []
  
  if (Array.isArray(obj.citations)) {
    for (const c of obj.citations) {
      if (!c || typeof c !== 'object') continue
      const cit = c as Record<string, unknown>
      const url = typeof cit.url === 'string' ? cit.url : null
      const domain = typeof cit.domain === 'string' ? cit.domain : null
      if (!url || !domain) continue
      const citation: any = {
        url,
        domain
      }
      if (typeof cit.entity_ref === 'string') {
        citation.entity_ref = cit.entity_ref
      }
      citations.push(citation)
    }
  }

  // Extract summary
  const summary = typeof obj.answer_summary === 'string'
    ? obj.answer_summary
    : typeof obj.summary === 'string'
    ? obj.summary
    : 'No summary provided.'

  // Extract flags
  const notesObj = obj.notes as Record<string, unknown> | undefined
  const flagsRaw = Array.isArray(notesObj?.flags) ? notesObj.flags : []
  const allowedFlags = new Set([
    'no_sources',
    'possible_hallucination',
    'outdated_info',
    'nap_mismatch',
    'conflicting_prices'
  ])
  const flags = flagsRaw.filter((f): f is string => 
    typeof f === 'string' && allowedFlags.has(f)
  ) as AnswerBlock['notes']['flags']

  const normalized = {
    ordered_entities: orderedEntities,
    citations,
    answer_summary: summary,
    notes: { flags }
  }

  const validated = AnswerBlockSchema.safeParse(normalized)
  return validated.success ? validated.data : null
}

export class ClaudeConnector implements Connector {
  surface = 'claude' as const

  async invoke(context: ConnectorContext): Promise<ConnectorResult> {
    const config = getGeoConfig()
    const client = new Anthropic({ 
      apiKey: config.anthropicApiKey,
      timeout: 600000, // 10 minutes timeout for slow API responses
    })
    const prompt = buildPrompt(context)
    const enableWebSearch = process.env.GEO_ENABLE_WEB_SEARCH === 'true'

    console.log('[claude] Query:', context.queryText)
    console.log('[claude] Brand:', context.brandName)
    if (enableWebSearch) {
      console.log('[claude] Web search: enabled')
    }

    let raw: unknown = null
    let parsed: AnswerBlock | null = null
    const messages: Anthropic.MessageParam[] = [
      {
        role: 'user',
        content: prompt
      }
    ]

    try {
      // First call - with web search tool if enabled
      const firstResponse = await client.messages.create({
        model: config.claudeModel,
        max_tokens: 2000,
        temperature: config.temperature,
        system: enableWebSearch
          ? 'You are a precise GEO audit assistant. Use web search to find current, accurate information about real estate properties. After searching, output strict JSON only.'
          : 'You are a precise GEO audit assistant. You must output ONLY valid JSON without any markdown formatting, code blocks, or explanatory text. Return raw JSON that can be directly parsed.',
        messages,
        ...(enableWebSearch ? {
          tools: [
            {
              name: 'web_search',
              description: 'Search the web for current information about real estate properties, reviews, listings, and competitive analysis',
              input_schema: {
                type: 'object',
                properties: {
                  query: {
                    type: 'string',
                    description: 'The search query to execute'
                  }
                },
                required: ['query']
              }
            }
          ]
        } : {})
      })

      raw = firstResponse

      // Check if Claude wants to use web search
      const toolUseBlock = firstResponse.content.find(
        (block): block is Anthropic.ToolUseBlock => block.type === 'tool_use'
      )

      if (toolUseBlock && toolUseBlock.name === 'web_search') {
        console.log('[claude] Tool use requested:', toolUseBlock.input)
        
        // Execute web search
        const searchQuery = (toolUseBlock.input as any).query || context.queryText
        const searchResults = await performWebSearch(searchQuery)
        const formattedResults = formatSearchResultsForLLM(searchResults)

        console.log('[claude] Search completed, ${searchResults.results.length} results')

        // Send search results back to Claude
        messages.push({
          role: 'assistant',
          content: firstResponse.content
        })
        messages.push({
          role: 'user',
          content: [
            {
              type: 'tool_result',
              tool_use_id: toolUseBlock.id,
              content: formattedResults
            },
            {
              type: 'text',
              text: 'Based on these search results, provide your answer in the required JSON format with ordered_entities and citations.'
            }
          ]
        })

        // Second call - get structured response
        const secondResponse = await client.messages.create({
          model: config.claudeModel,
          max_tokens: 1200,
          temperature: config.temperature,
          system: 'You are a precise GEO audit assistant. Output ONLY valid JSON without any markdown formatting. Return raw JSON that can be directly parsed.',
          messages
        })

        const textBlocks = (secondResponse.content ?? []).filter(
          (b): b is Anthropic.TextBlock => b.type === 'text'
        )
        const contentText = textBlocks.map(b => b.text).join('\n')

        const jsonValue = tryParseJson(contentText)
        if (jsonValue) {
          parsed = coerceToAnswerBlock(jsonValue)
          if (parsed) {
            console.log('[claude] ✓ Parsed', parsed.ordered_entities.length, 'entities,', parsed.citations.length, 'citations (with search)')
          }
        }
      } else {
        // No tool use - direct response
        const textBlocks = (firstResponse.content ?? []).filter(
          (b): b is Anthropic.TextBlock => b.type === 'text'
        )
        const contentText = textBlocks.map(b => b.text).join('\n')

        const jsonValue = tryParseJson(contentText)
        if (jsonValue) {
          parsed = coerceToAnswerBlock(jsonValue)
          if (parsed) {
            console.log('[claude] ✓ Parsed', parsed.ordered_entities.length, 'entities,', parsed.citations.length, 'citations')
          }
        }
      }
    } catch (error) {
      console.error('[claude] API error:', error)
      raw = { error: error instanceof Error ? error.message : String(error) }
    }

    if (!parsed) {
      console.warn('[claude] Returning fallback answer')
      return {
        answer: {
          ordered_entities: [],
          citations: [],
          answer_summary: 'No structured sources returned',
          notes: { flags: ['no_sources'] }
        },
        raw
      }
    }

    return { answer: parsed, raw }
  }
}

File: p11-platform/services/data-engine/connectors/claude_connector.py
"""
Claude (Anthropic) Connector for PropertyAudit (Structured Mode)
Full parity with TypeScript claude-connector.ts including:
- Location context
- Quality flags
- Retry logic
"""
import os
import logging
from typing import Dict, Any, Optional
import anthropic
from tenacity import retry, stop_after_attempt, wait_exponential

logger = logging.getLogger(__name__)


def build_prompt(context: Dict[str, Any]) -> str:
    """Build detailed prompt matching TypeScript version."""
    domains = ', '.join(context.get('brandDomains', []))
    competitors = ', '.join(context.get('competitors', []))
    location = context.get('propertyLocation', {})
    
    lines = [
        "Task: Perform a GEO audit for this specific apartment property and return ONLY the JSON object matching the schema."
    ]
    
    # Add property location context
    if location and location.get('city') and location.get('state'):
        lines.extend([
            "",
            "Property Details:",
            f"- Name: {context['brandName']}",
            f"- Location: {location['city']}, {location['state']}"
        ])
        
        if location.get('fullAddress'):
            lines.append(f"- Address: {location['fullAddress']}")
        if location.get('websiteUrl'):
            lines.append(f"- Official Website: {location['websiteUrl']}")
        
        lines.extend([
            "",
            f"CRITICAL: This property is located in {location['city']}, {location['state']}.",
            "Do NOT confuse with properties in other cities or states.",
            f"Verify all information relates to the {location['city']}, {location['state']} location."
        ])
    
    lines.extend([
        "",
        f"Query: {context['queryText']}",
        f"Brand: {context['brandName']}",
        f"Brand domains: {domains or '—'}",
        f"Competitors: {competitors or '—'}",
        "Requirements:",
        "- Produce an ordered list of providers/brands relevant to the query (name, domain, rationale, position starting at 1).",
        "- Include citations with absolute URLs and their domains.",
        "- Summarize the answer in 1-2 sentences.",
        "- If no grounded sources are available, set notes.flags to include \"no_sources\".",
        "Output: Return ONLY the JSON object, no markdown, no explanations."
    ])
    
    return '\n'.join(lines)


def coerce_to_answer_block(candidate: Any) -> Optional[Dict[str, Any]]:
    """Coerce response to standard AnswerBlock format."""
    if not candidate or not isinstance(candidate, dict):
        return None
    
    entities_source = (
        candidate.get('ordered_entities') or
        candidate.get('results') or
        candidate.get('providers')
    )
    
    if not entities_source or not isinstance(entities_source, list):
        return None
    
    ordered_entities = []
    for idx, item in enumerate(entities_source):
        if not isinstance(item, dict):
            continue
        
        name = item.get('name')
        domain = item.get('domain')
        
        if not name or not domain:
            continue
        
        ordered_entities.append({
            'name': str(name),
            'domain': str(domain),
            'rationale': str(item.get('rationale', 'No rationale provided.')),
            'position': int(item.get('position', idx + 1))
        })
    
    if not ordered_entities:
        return None
    
    citations_source = candidate.get('citations', [])
    citations = []
    for c in citations_source:
        if not isinstance(c, dict):
            continue
        url = c.get('url')
        domain = c.get('domain')
        if url and domain:
            citations.append({
                'url': str(url),
                'domain': str(domain),
                'entity_ref': c.get('entity_ref')
            })
    
    summary = (
        candidate.get('answer_summary') or
        candidate.get('summary') or
        'No summary provided.'
    )
    
    notes = candidate.get('notes', {})
    flags = notes.get('flags', []) if isinstance(notes, dict) else []
    
    # Add quality flags
    if not citations:
        flags.append('no_sources')
    
    allowed_flags = {'no_sources', 'possible_hallucination', 'outdated_info', 'nap_mismatch', 'conflicting_prices'}
    flags = [f for f in flags if f in allowed_flags]
    
    return {
        'ordered_entities': ordered_entities,
        'citations': citations,
        'answer_summary': str(summary),
        'notes': {'flags': flags}
    }


class ClaudeConnector:
    """Claude connector with full feature parity to TypeScript."""
    
    def __init__(self):
        self.api_key = os.environ.get('ANTHROPIC_API_KEY')
        if not self.api_key:
            raise ValueError("ANTHROPIC_API_KEY not set")
        
        self.client = anthropic.Anthropic(api_key=self.api_key)
        self.model = os.environ.get('GEO_CLAUDE_MODEL', 'claude-sonnet-5')
    
    @retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=4, max=10))
    async def invoke(self, context: Dict[str, Any]) -> Dict[str, Any]:
        """
        Invoke Claude with full feature set.
        
        Args:
            context: Query context with brandName, brandDomains, queryText, etc.
            
        Returns:
            Dict with 'answer' and 'raw' response
        """
        query_text = context['queryText']
        
        logger.info(f"[Claude] Query: {query_text[:50]}... | Model: {self.model}")
        
        prompt = build_prompt(context)
        
        try:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=2000,
                temperature=0.3,
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )
            
            content = response.content[0].text
            logger.debug(f"[Claude] Response: {len(content)} chars")
            
            # Parse JSON
            import json
            import re
            try:
                parsed = json.loads(content)
            except json.JSONDecodeError:
                # Try to extract JSON
                match = re.search(r'\{[\s\S]*\}', content)
                if match:
                    parsed = json.loads(match.group(0))
                else:
                    raise ValueError("Could not parse JSON from response")
            
            # Coerce to standard format
            answer = coerce_to_answer_block(parsed)
            
            if not answer:
                logger.warning("[Claude] Failed to coerce response to AnswerBlock format")
                answer = {
                    'ordered_entities': [],
                    'citations': [],
                    'answer_summary': content[:200] if content else 'No response',
                    'notes': {'flags': ['possible_hallucination']}
                }
            
            return {
                'answer': answer,
                'raw': {
                    'model': self.model,
                    'response_id': response.id,
                    'usage': {
                        'input_tokens': response.usage.input_tokens,
                        'output_tokens': response.usage.output_tokens
                    },
                    'stop_reason': response.stop_reason
                }
            }
            
        except Exception as e:
            logger.error(f"[Claude] Error: {e}", exc_info=True)
            return {
                'answer': {
                    'ordered_entities': [],
                    'citations': [],
                    'answer_summary': f'Error: {str(e)}',
                    'notes': {'flags': ['no_sources']}
                },
                'raw': {'error': str(e), 'model': self.model}
            }






